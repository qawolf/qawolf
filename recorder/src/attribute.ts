export type AttributeValuePair = {
  name: string;
  value: string;
};

type GetAttribueValue = {
  attribute: string;
  element: HTMLElement;
};

type GetRegexAttribute = {
  element: HTMLElement;
  regexString: string;
};

export const deserializeRegex = (regexString: string): RegExp | null => {
  try {
    const parts = regexString.match(/\/(.*)\/(.*)?/) as RegExpMatchArray;

    return new RegExp(parts[1], parts[2] || "");
  } catch (e) {
    console.error(
      `qawolf: invalid regex attribute ${regexString}, skipping this attribute`
    );
    return null;
  }
};

export const getRegexAttribute = ({
  element,
  regexString,
}: GetRegexAttribute): AttributeValuePair | null => {
  const regex = deserializeRegex(regexString);
  if (!regex) return null;

  const attributes = element.attributes;

  for (let i = 0; i < attributes.length; i++) {
    const { name, value } = attributes[i];

    if (name.match(regex)) {
      return { name, value };
    }
  }

  return null;
};

export const getAttribute = ({
  attribute,
  element,
}: GetAttribueValue): AttributeValuePair | null => {
  const isRegex = attribute[0] === "/";

  if (isRegex) {
    return getRegexAttribute({
      element,
      regexString: attribute,
    });
  }

  const value = element.getAttribute(attribute);
  if (!value) return null;

  return { name: attribute, value };
};

export const hasAttribute = (
  element: HTMLElement,
  attributes: string[]
): boolean => {
  return !!attributes.find((attribute) => getAttribute({ attribute, element }));
};
