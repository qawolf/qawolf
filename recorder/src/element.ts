import { getXpath } from "./serialize";

type OnElementAddedToGroupFn = (element: HTMLElement, depth: number) => void;

type TraverseClickableElementsInput = {
  ancestorChain?: string[];
  depth?: number;
  direction?: "up" | "down";
  element: HTMLElement;
  group: HTMLElement[];
  maxDepth?: number;
  onElementAddedToGroup?: OnElementAddedToGroupFn;
};

const BUTTON_INPUT_TYPES = ["button", "image", "reset", "submit"];
const CLICK_GROUP_ELEMENTS = ["a", "button", "label"];
const MAX_CLICKABLE_ELEMENT_TRAVERSE_DEPTH = 10;

export const isVisible = (
  element: Element,
  computedStyle?: CSSStyleDeclaration
): boolean => {
  const htmlElement = element as HTMLElement;
  if (htmlElement.offsetWidth <= 0 || htmlElement.offsetHeight <= 0) {
    return false;
  }

  if (computedStyle && computedStyle.visibility === "hidden") {
    return false;
  }

  if (computedStyle && computedStyle.display === "none") {
    return false;
  }

  return true;
};

export const isLikelyTopOfClickGroup = (element: HTMLElement): boolean => {
  const tagName = element.tagName.toLowerCase();
  return (
    CLICK_GROUP_ELEMENTS.includes(tagName) ||
    element.getAttribute("role") === "button" ||
    (tagName === "input" &&
      BUTTON_INPUT_TYPES.includes(element.getAttribute("type")))
  );
};

/**
 * @summary Traverse the DOM in both directions, adding clickable elements to the array
 *   until we've determined the full set of elements that are in the clickable area.
 *   This is not foolproof because we can't know where exactly click handlers might
 *   be attached, but we can do a pretty good job of guessing.
 */
const traverseClickableElements = (
  input: TraverseClickableElementsInput
): void => {
  const {
    ancestorChain = [],
    depth = 0,
    direction = "up",
    element,
    group,
    maxDepth = MAX_CLICKABLE_ELEMENT_TRAVERSE_DEPTH,
    onElementAddedToGroup,
  } = input;

  // Regardless of which direction we're moving, stop if we hit an invisible element
  if (!isVisible(element, window.getComputedStyle(element))) return;

  // When moving up, when we reach the topmost clickable element, we
  // stop traversing up and begin traversing down from there.
  if (direction === "up" && isLikelyTopOfClickGroup(element)) {
    traverseClickableElements({
      direction: "down",
      element,
      group,
      maxDepth,
      onElementAddedToGroup,
    });
    return;
  }

  // When moving down, stop if we hit the top of another click group (nested buttons)
  if (direction === "down" && depth > 0 && isLikelyTopOfClickGroup(element))
    return;

  const newDepth = depth + 1;
  const lowerTagName = element.tagName.toLowerCase();

  if (direction === "up") {
    // Call self for the parent element, incrementing depth
    if (element.parentElement) {
      traverseClickableElements({
        ancestorChain: [lowerTagName, ...ancestorChain],
        direction,
        element: element.parentElement,
        group,
        maxDepth,
        depth: newDepth,
        onElementAddedToGroup,
      });
    }
  } else {
    // Respect max depth only when going down
    if (newDepth > maxDepth) return;

    // If we make it this far, this element should be part of the current group.
    // We add elements to the group only on the way down to avoid adding any twice.
    group.push(element);

    // Let caller do additional things with each element as we add it
    if (onElementAddedToGroup) onElementAddedToGroup(element, depth);

    const newAncestorChain = [...ancestorChain, lowerTagName];
    console.debug(
      "qawolf: added %s to click group",
      newAncestorChain.join(" > ")
    );

    // For now, let's skip going down into SVG descendants to keep this fast. If anyone
    // finds a situation in which this causes problems, we can remove this.
    if (lowerTagName !== "svg") {
      for (const child of element.children) {
        // Call self for each child element, incrementing depth
        traverseClickableElements({
          ancestorChain: newAncestorChain,
          direction,
          element: child as HTMLElement,
          group,
          maxDepth,
          depth: newDepth,
          onElementAddedToGroup,
        });
      }
    }
  }
};

/**
 * @summary Sometimes there is a group of elements that together make up what appears
 *   to be a single button, link, image, etc. Examples: a > span | button > div > span
 *   For these we want to take into consideration the entire "clickable group" when
 *   building a good selector. The topmost clickable (a | button | input) should be
 *   preferred in many cases, but if an inner element has a lower-penalty attribute
 *   then that should be preferred.
 *
 * @return An array of HTMLElement that make up the clickable group. If `element`
 *   itself is not clickable, the array is empty.
 */
export const getClickableGroup = (
  element: HTMLElement,
  onElementAddedToGroup?: OnElementAddedToGroupFn
): HTMLElement[] => {
  console.debug("qawolf: get clickable ancestor for", getXpath(element));

  const group = [];

  // Recursive function that will mutate clickableElements array. A recursive
  // function is better than loops to avoid blocking UI paint.
  traverseClickableElements({ element, group, onElementAddedToGroup });

  return group;
};

/**
 * @summary Returns the topmost isContentEditable ancestor. Editable areas can
 *   have HTML elements in them, and these elements emit events, but in general
 *   I don't think we want to keep track of anything within the editable area.
 *   For example, if you click a particular paragraph in a `contenteditable`
 *   div, we should just record it as a click/focus of the editable div.
 */
export const getTopmostEditableElement = (
  element: HTMLElement
): HTMLElement => {
  if (!element.isContentEditable) return element;

  console.debug("qawolf: get editable ancestor for", getXpath(element));

  let ancestor = element;
  do {
    if (!ancestor.parentElement || !ancestor.parentElement.isContentEditable) {
      console.debug(
        `qawolf: found editable ancestor: ${ancestor.tagName}`,
        getXpath(ancestor)
      );
      return ancestor;
    }

    ancestor = ancestor.parentElement;
  } while (ancestor);

  // This should never be hit, but here as a safety
  return element;
};

/**
 * @summary Returns the current "value" of an element. Pass in an event `target`.
 *   For example, returns the `.value` or the `.innerText` of a content-editable.
 *   If no value can be determined, returns `null`.
 */
export const getInputElementValue = (
  element: HTMLInputElement
): string | null => {
  let inputValue: string;

  // In the wild, we've seen examples of input elements with `contenteditable=true`,
  // but an `input` never has inner text, so we check for `input` tag name here.
  if (element.isContentEditable && element.tagName.toLowerCase() !== "input") {
    inputValue = element.innerText;
  } else {
    inputValue = element.value;
  }

  return typeof inputValue === "string" ? inputValue : null;
};
