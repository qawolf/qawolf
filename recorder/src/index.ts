export * from "./qawolf";

import * as qawolf from "./qawolf";

export type QAWolfWeb = typeof qawolf;
