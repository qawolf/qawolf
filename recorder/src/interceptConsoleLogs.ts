import { getXpath } from "./serialize";
import { Callback, LogEvent } from "./types";

const LOG_LEVELS = ["debug", "error", "info", "log", "warn"];

type LogEventCallback = Callback<LogEvent>;

export const formatArgument = (argument: unknown): string => {
  if (typeof argument === "string") {
    return argument;
  }

  if (argument && (argument as any).nodeName) {
    // log nodes as their xpath
    return getXpath(argument as Node);
  }

  try {
    return JSON.stringify(argument);
  } catch (error) {
    if (argument && argument.toString) {
      return argument.toString();
    }

    return "";
  }
};

export const interceptConsoleLogs = (): void => {
  if ((window as any).qawInterceptLogs) return;
  (window as any).qawInterceptLogs = true;

  LOG_LEVELS.forEach((level: keyof Console) => {
    const browserLog = console[level].bind(console);

    console[level] = (...args: any): void => {
      const message: string = args
        .map((arg: any) => formatArgument(arg))
        .join(" ");

      browserLog(...args);

      // playwright is currently using logs as its message transport. prevent infinite loop.
      if (message.includes("__playwright")) return;

      const logCallback: LogEventCallback = (window as any).qawLogEvent;
      if (logCallback) logCallback({ level, message });
    };
  });
};
