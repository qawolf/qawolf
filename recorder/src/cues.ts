import { getAttribute } from "./attribute";
import { buildTextSelector } from "./selectorEngine";
import { getValueMatches } from "./isDynamic";

const DYNAMIC_VALUE_OK_ATTRIBUTES = ["placeholder", "href", "src", "value"];

export type Cue = {
  level: number; // 0 is target, 1 is parent, etc.
  penalty: number; // Cue type penalty plus PENALTY_PER_LEVEL
  type: string;
  value: string;
};

export type BuildCues = {
  cueTypesConfig: CueTypesConfig;
  isClick: boolean;
  target: HTMLElement;
};

type CueTypeConfig = {
  elements: string[];
  isPreferred?: boolean;
  penalty: number;
};

type CueTypesConfig = Record<string, CueTypeConfig>;

type BuildCuesForElement = {
  cueTypesConfig: CueTypesConfig;
  element: HTMLElement;
  isClick: boolean;
  level: number;
};

/**
 * For each cue type, this defines which elements it applies to and a penalty
 * value. Higher penalty means it is less likely that this will be helpful in
 * constructing a unique, short, and attractive selector. We can and should
 * work to fine tune the penalty values over time.
 *
 * Users may add attributes to this list using the `attribute` config option.
 */
const ConfigByCueType: CueTypesConfig = {
  alt: {
    elements: ["area", "img", "input[type=image]"],
    penalty: 20,
  },
  "aria-label": {
    elements: ["*"],
    penalty: 8,
  },
  class: {
    elements: ["*"],
    penalty: 10,
  },
  contenteditable: {
    elements: ["*"],
    // High penalty because it is unlikely to be unique given that the value is always the same
    penalty: 30,
  },
  for: {
    elements: ["label", "output"],
    penalty: 5,
  },
  href: {
    elements: ["a"],
    penalty: 15,
  },
  id: {
    elements: ["*"],
    penalty: 5,
  },
  name: {
    elements: [
      "button",
      "form",
      "fieldset",
      "iframe",
      "input",
      "object",
      "output",
      "select",
      "textarea",
      "map",
    ],
    penalty: 10,
  },
  placeholder: {
    elements: ["input", "textarea"],
    penalty: 12,
  },
  src: {
    elements: ["audio", "iframe", "img", "input[type=image]", "video"],
    penalty: 15,
  },
  tag: {
    elements: ["*"],
    penalty: 40,
  },
  text: {
    elements: ["*"],
    penalty: 12,
  },
  title: {
    elements: ["*"],
    penalty: 10,
  },
  value: {
    elements: [
      "option",
      "button",
      "input[type=submit]",
      "input[type=button]",
      "input[type=checkbox]",
      "input[type=radio]",
    ],
    penalty: 10,
  },
};

/**
 * @summary Get final cue types config
 * @return Cue type config with preferred attributes added. For now,
 *   all attributes are given 0 penalty, but eventually the user
 *   config could support custom penalties for each.
 */
export const getCueTypesConfig = (attributes: string[]): CueTypesConfig => {
  const preferredAttributes: CueTypesConfig = {};

  attributes.forEach((attribute) => {
    preferredAttributes[attribute] = {
      elements: ["*"],
      isPreferred: true,
      penalty: 0,
    };
  }, {});

  return {
    ...ConfigByCueType,
    ...preferredAttributes,
  };
};

export const buildCueValueForTag = (element: HTMLElement): string => {
  const tagName = element.tagName.toLowerCase();
  if (!element.parentElement) return tagName;

  const siblings = element.parentElement.children;
  const sameTagSiblings: HTMLElement[] = [];

  for (const sibling of siblings) {
    if (sibling.tagName.toLowerCase() === tagName) {
      sameTagSiblings.push(sibling as HTMLElement);
    }
  }

  if (sameTagSiblings.length < 2) {
    return tagName;
  }

  const nthIndex = sameTagSiblings.indexOf(element) + 1;
  if (nthIndex === 1) return tagName;

  return `${tagName}:nth-of-type(${nthIndex})`;
};

export const buildCuesForElement = ({
  cueTypesConfig,
  element,
  isClick,
  level,
}: BuildCuesForElement): Cue[] => {
  // For body and html, we never have more than one, so
  // just 'tag' cue is needed and we can save some time.
  const tagName = element.tagName.toLowerCase();
  if (["html", "body"].includes(tagName)) {
    return [
      {
        level,
        penalty: ConfigByCueType.tag.penalty,
        type: "tag",
        value: tagName,
      },
    ];
  }

  const cues: Cue[] = Object.keys(cueTypesConfig).reduce((list, cueType) => {
    const { elements, isPreferred, penalty } = cueTypesConfig[cueType];

    // First find out whether this cue type is relevant for this element
    if (!elements.some((selector: string) => element.matches(selector))) {
      return list;
    }

    switch (cueType) {
      // Special handling for "class" attribute
      case "class": {
        element.classList.forEach((c) => {
          // Get all exact and partial matches for this one class.
          // However, all classes are space-delimited within a single
          // class attribute, so we need to change starts-with and
          // ends-with attribute operators to `contains`
          getValueMatches(c).forEach(({ match, operator }) => {
            if (operator === "=") {
              list.push({
                level,
                penalty,
                type: "class",
                value: `.${CSS.escape(c)}`,
              });
            } else {
              list.push({
                level,
                penalty: penalty + 5, // add penalty because it's a partial match
                type: "attribute",
                value: `[class*="${match}"]`,
              });
            }
          });
        });
        break;
      }
      // Special handling for "id" attribute
      case "id": {
        getValueMatches(element.id).forEach(({ match, operator }) => {
          // This is mostly the same as the attribute logic below except
          // that we use `#id` notation in place of an `=` attribute selector.
          if (operator === "=") {
            list.push({
              level,
              penalty,
              type: "id",
              value: `#${CSS.escape(match)}`,
            });
          } else {
            list.push({
              level,
              penalty: penalty + 5, // add penalty because it's a partial match
              type: "attribute",
              value: `[id${operator}"${match}"]`,
            });
          }
        });
        break;
      }
      // Special handling for "tag" type
      case "tag":
        list.push({
          level,
          penalty,
          type: "tag",
          value: buildCueValueForTag(element),
        });
        break;
      // Special handling for "text" type
      case "text": {
        if (!isClick) return list;
        // Generate text cues only for level 0 (the actual click target, or targets
        // in the case of a click group). It's unlikely that we would choose a text
        // selector for any element that is higher, and building a text selector becomes
        // extremely slow when some of the ancestor elements have many descendants.
        if (level > 0) return list;
        const value = buildTextSelector(element);
        if (typeof value === "string") {
          list.push({ level, penalty, type: "text", value });
        }
        break;
      }
      // Everything else is just an attribute
      default: {
        const attributeValuePair = getAttribute({
          attribute: cueType,
          element,
        });
        if (attributeValuePair) {
          const { name, value } = attributeValuePair;
          if (
            value.length &&
            (isPreferred || DYNAMIC_VALUE_OK_ATTRIBUTES.includes(name))
          ) {
            list.push({
              level,
              penalty,
              type: "attribute",
              value: `[${name}="${value}"]`,
            });
          } else {
            getValueMatches(value).forEach(({ match, operator }) => {
              list.push({
                level,
                penalty,
                type: "attribute",
                value: `[${name}${operator}"${match}"]`,
              });
            });
          }
        }
        break;
      }
    }

    return list;
  }, []);

  return cues;
};

export const buildCues = ({
  cueTypesConfig,
  isClick,
  target,
}: BuildCues): Cue[] => {
  const cues: Cue[] = [];
  let element: HTMLElement = target;
  let level = 0;

  while (element) {
    cues.push(
      ...buildCuesForElement({ cueTypesConfig, element, isClick, level })
    );

    element = element.parentElement;
    level += 1;
  }

  return cues;
};

export const findNearestPreferredAttributeCue = (cues: Cue[]): Cue | null => {
  return cues.reduce((foundCue, cue) => {
    if (cue.penalty === 0 && (!foundCue || foundCue.level > cue.level)) {
      return cue;
    }

    return foundCue;
  }, null);
};

export const getPenalty = (cues: Cue[]): number => {
  return cues.reduce((a, b) => a + b.penalty, 0);
};

export const getValueLength = (cues: Cue[]): number => {
  return cues.reduce((total, cue) => total + cue.value.length, 0);
};
