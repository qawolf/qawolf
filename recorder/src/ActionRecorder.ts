import { resolveAction } from "./resolveAction";
import { getInputElementValue } from "./element";
import { buildSelector } from "./selector";
import { Action, Callback, ElementAction, PossibleAction } from "./types";

type ActionCallback = Callback<ElementAction>;

type ConstructorOptions = {
  attribute?: string;
};

export class ActionRecorder {
  private _attributes: string[];
  private _lastReceivedAction: PossibleAction;
  private _lastRecordedAction: ElementAction;
  private _onDispose: Callback[] = [];

  constructor(options: ConstructorOptions = {}) {
    this._attributes = (options.attribute || "").split(",");
    console.debug("ActionRecorder: created", options);
    this.start();
  }

  public stop(): void {
    this._onDispose.forEach((d) => d());
    console.debug("ActionRecorder: stopped");
  }

  private listen(
    eventName: "click" | "input" | "change" | "keydown",
    handler: (ev: MouseEvent | KeyboardEvent | Event) => any
  ): void {
    document.addEventListener(eventName, handler, {
      capture: true,
      passive: true,
    });

    this._onDispose.push(() =>
      document.removeEventListener(eventName, handler, {
        // `capture` value must match for proper removal
        // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener#Matching_event_listeners_for_removal
        capture: true,
      })
    );
  }

  private recordAction(
    action: Action,
    event: MouseEvent | KeyboardEvent | Event,
    value?: string
  ): void {
    console.debug(`ActionRecorder: ${action} action detected`, event);

    const actionCallback: ActionCallback = (window as any).qawElementAction;
    if (!actionCallback) {
      console.debug(
        "ActionRecorder: can't record actions without an action callback"
      );
      return;
    }

    const time = Date.now();

    // This is mainly to make resolveAction easier to test.
    // We then don't need to mock complicated browser Event object.
    const possibleAction: PossibleAction = {
      action,
      isTrusted: event.isTrusted,
      target: event.target as HTMLElement,
      time,
      value,
    };

    action = resolveAction(possibleAction, this._lastReceivedAction);

    this._lastReceivedAction = possibleAction;

    // If no action was returned, this isn't an event we care about
    // so we can skip building a selector and emitting it.
    if (!action) return;

    const selector = buildSelector({
      attributes: this._attributes,
      isClick: action === "click",
      target: event.target as HTMLElement,
    });

    const elementAction: ElementAction = {
      action,
      selector,
      time,
    };

    if (value !== undefined) {
      // Value should be coerced to an empty string
      elementAction.value = typeof value === "string" ? value : "";
    }

    // Fills come from both "input" and "change" events for completeness, but this
    // means that we could end up emitting back-to-back fills with the same value.
    // We can check here to avoid that. (For press and click, back-to-back identical
    // events could be valid.)
    if (
      ((action === "fill" && this._lastRecordedAction.action === "fill") ||
        (action === "selectOption" &&
          this._lastRecordedAction.action === "selectOption")) &&
      elementAction.selector === this._lastRecordedAction.selector &&
      elementAction.value === this._lastRecordedAction.value
    ) {
      console.debug(`ActionRecorder: skipping duplicate ${action}`);
      return;
    }

    console.debug(`ActionRecorder: ${action} action recorded:`, elementAction);

    this._lastRecordedAction = elementAction;

    actionCallback(elementAction);
  }

  public start(): void {
    //////// MOUSE EVENTS ////////

    this.listen("click", (event) => {
      // only the main button (not right clicks/etc)
      // https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
      if ((event as MouseEvent).button !== 0) return;

      this.recordAction("click", event);
    });

    //////// INPUT EVENTS ////////

    this.listen("input", (event) => {
      const target = event.target as HTMLInputElement;
      this.recordAction("fill", event, getInputElementValue(target));
    });

    this.listen("change", (event) => {
      const target = event.target as HTMLInputElement;
      this.recordAction("fill", event, getInputElementValue(target));
    });

    //////// KEYBOARD EVENTS ////////

    this.listen("keydown", (event) => {
      this.recordAction("press", event, (event as KeyboardEvent).key);
    });

    console.debug("ActionRecorder: started");
  }
}
