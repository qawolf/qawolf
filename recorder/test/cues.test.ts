import { Browser, Page } from "playwright";
import { QAWolfWeb } from "../src";
import { Cue } from "../src/cues";
import { DEFAULT_ATTRIBUTE_LIST, launch, TEST_URL } from "./utils";

let browser: Browser;
let page: Page;

beforeAll(async () => {
  browser = await launch();
  const context = await browser.newContext();
  page = await context.newPage();
  await page.goto(`${TEST_URL}checkbox-inputs`);
});

afterAll(() => browser.close());

describe("buildCues", () => {
  const buildCues = async (selector: string): Promise<Cue[]> => {
    return page.evaluate(
      ({ attributes, selector }) => {
        const qawolf: QAWolfWeb = (window as any).qawolf;
        const target = document.querySelector(selector) as HTMLElement;
        const cueTypesConfig = qawolf.getCueTypesConfig(attributes);

        return qawolf.buildCues({
          cueTypesConfig,
          isClick: true,
          target,
        });
      },
      { attributes: DEFAULT_ATTRIBUTE_LIST.split(","), selector }
    );
  };

  it("builds cues for a target", async () => {
    const cues = await buildCues("#single");
    expect(cues).toMatchSnapshot();
  });
});

describe("buildCuesForElement", () => {
  const buildCuesForElement = async (selector: string, level = 1): Promise<Cue[]> => {
    return page.evaluate(
      ({ level, selector }) => {
        const qawolf: QAWolfWeb = (window as any).qawolf;
        const element = document.querySelector(selector) as HTMLElement;
        if (!element) return [];

        const cueTypesConfig = qawolf.getCueTypesConfig(["data-qa"]);
        return qawolf.buildCuesForElement({
          cueTypesConfig,
          element,
          isClick: true,
          level,
        });
      },
      {
        level,
        selector,
      }
    );
  };

  it("builds cues for an element", async () => {
    const cues = await buildCuesForElement("#single");
    expect(cues).toEqual([
      { level: 1, penalty: 5, type: "id", value: "#single" },
      { level: 1, penalty: 40, type: "tag", value: "input" },
      {
        level: 1,
        penalty: 0,
        type: "attribute",
        value: '[data-qa="html-checkbox"]',
      },
    ]);

    const cues2 = await buildCuesForElement('[for="single"]');
    expect(cues2).toEqual([
      { level: 1, penalty: 5, type: "attribute", value: '[for="single"]' },
      { level: 1, penalty: 40, type: "tag", value: "label" },
    ]);

    const cues3 = await buildCuesForElement("#special\\:id");
    expect(cues3).toEqual([
      {
        level: 1,
        penalty: 10,
        type: "class",
        value: ".special\\:class",
      },
      {
        level: 1,
        penalty: 5,
        type: "id",
        value: "#special\\:id",
      },
      {
        level: 1,
        penalty: 40,
        type: "tag",
        value: "input:nth-of-type(2)",
      },
    ]);
  });

  it("builds text cues for level 0 only", async () => {
    const cues = await buildCuesForElement('[for="single"]', 0);
    expect(cues).toEqual([
      { level: 0, penalty: 5, type: "attribute", value: '[for="single"]' },
      { level: 0, penalty: 40, type: "tag", value: "label" },
      { level: 0, penalty: 12, type: "text", value: '" Single checkbox"' },
    ]);

    const cues2 = await buildCuesForElement('[for="single"]', 1);
    expect(cues2).toEqual([
      { level: 1, penalty: 5, type: "attribute", value: '[for="single"]' },
      { level: 1, penalty: 40, type: "tag", value: "label" },
    ]);
  });

  it("ignores dynamic attr values when building cues", async () => {
    const cues = await buildCuesForElement('[name="bu32879fDi"]');
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(4)",
        },
      ]
    `);
  });

  it("ignores empty attr values when building cues", async () => {
    const cues = await buildCuesForElement("#y908drgun4");
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(8)",
        },
      ]
    `);
  });

  it("ignores empty text values when building cues", async () => {
    await page.setContent(`
      <html>
        <body>
          <header>&nbsp;</header>
        </body>
      </html>
    `);
    const cues = await buildCuesForElement("header");
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "header",
        },
      ]
    `);
    await page.goto(`${TEST_URL}checkbox-inputs`);
  });

  it("uses non-dynamic attr values when building cues", async () => {
    const cues = await buildCuesForElement('[name="nonDynamicInput"]');
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 10,
          "type": "attribute",
          "value": "[name=\\"nonDynamicInput\\"]",
        },
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(3)",
        },
      ]
    `);
  });

  it("builds cues with attributes that have dynamic beginning", async () => {
    const cues = await buildCuesForElement("#ern84j8g0");
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 10,
          "type": "attribute",
          "value": "[name$=\\"-check\\"]",
        },
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(6)",
        },
      ]
    `);
  });

  it("builds cues with attributes that have dynamic ending", async () => {
    const cues = await buildCuesForElement("#v9eonirh894");
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 10,
          "type": "attribute",
          "value": "[name^=\\"input-\\"]",
        },
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(5)",
        },
      ]
    `);
  });

  it("builds cues with attributes that have dynamic beginning and ending", async () => {
    const cues = await buildCuesForElement("#fdg8e9v4");
    expect(cues).toMatchInlineSnapshot(`
      Array [
        Object {
          "level": 1,
          "penalty": 10,
          "type": "attribute",
          "value": "[name*=\\"-blue-\\"]",
        },
        Object {
          "level": 1,
          "penalty": 40,
          "type": "tag",
          "value": "input:nth-of-type(7)",
        },
      ]
    `);
  });
});

describe("buildCueValueForTag", () => {
  const buildCueValueForTag = async (selector: string): Promise<string> => {
    return page.evaluate((selector) => {
      const qawolf: QAWolfWeb = (window as any).qawolf;
      const element = document.querySelector(selector) as HTMLElement;

      return element
        ? qawolf.buildCueValueForTag(element)
        : "ELEMENT NOT FOUND";
    }, selector);
  };

  it("returns tag name if no parent element", async () => {
    const value = await buildCueValueForTag("html");
    expect(value).toBe("html");
  });

  it("returns tag name if element has no siblings", async () => {
    const value = await buildCueValueForTag(".container h3");
    expect(value).toBe("h3");
  });

  it("returns tag name if element is first child", async () => {
    const value = await buildCueValueForTag('[for="single"]');
    expect(value).toBe("label");
  });

  it("returns nth-of-type tag if element is a lower sibling", async () => {
    const value = await buildCueValueForTag('[for="special\\:id"]');
    expect(value).toBe("label:nth-of-type(2)");
  });
});
