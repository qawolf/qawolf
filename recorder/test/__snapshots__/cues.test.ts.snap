// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCues builds cues for a target 1`] = `
Array [
  Object {
    "level": 0,
    "penalty": 5,
    "type": "id",
    "value": "#single",
  },
  Object {
    "level": 0,
    "penalty": 40,
    "type": "tag",
    "value": "input",
  },
  Object {
    "level": 0,
    "penalty": 0,
    "type": "attribute",
    "value": "[data-qa=\\"html-checkbox\\"]",
  },
  Object {
    "level": 1,
    "penalty": 10,
    "type": "class",
    "value": ".container",
  },
  Object {
    "level": 1,
    "penalty": 40,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 2,
    "penalty": 40,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 3,
    "penalty": 5,
    "type": "id",
    "value": "#root",
  },
  Object {
    "level": 3,
    "penalty": 40,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 4,
    "penalty": 40,
    "type": "tag",
    "value": "body",
  },
  Object {
    "level": 5,
    "penalty": 40,
    "type": "tag",
    "value": "html",
  },
]
`;
